src/qdwh.cc

```diff
@@ -0,0 +1,410 @@
+            auto U = slate::HermitianMatrix<scalar_t>(
+                    slate::Uplo::Upper, W2 );
+            posv(U, AT, opts);
+
+            // Update A = (b/c) * A + (a-b/c) * ( linsolve( C, linsolve( C, A') ) )';
+            alpha = (a-b/c); beta = (b/c);
+            auto AT2 = conj_transpose(AT);
+            add(alpha, AT2, beta, A, opts);
+
+            itpo += 1;
+
+            //facto = 1;
+            // Main flops used in this step
+            //flops_dgemm  = FLOPS_DGEMM( m, n, n );
+            //flops_dpotrf = FLOPS_DPOTRF( m );
+            //flops_dtrsm  = FLOPS_DTRSM( 'L', m, n );
+            //flops += flops_dgemm + flops_dpotrf + 2. * flops_dtrsm;
+        }
+
+        // Check if it converge, compute the norm of matrix A - W1
+        conv = 10.0;
+        if (it >= itconv) {
+            add(one, A, -one, W10, opts);
+            conv = norm(slate::Norm::Fro, W10, opts);
+        }
+    }
+
+    if (A.mpiRank() == 0) {
+        printf("%-7s  %-7s\n", "QR", "PO");
+        printf("%-7d  %-7d\n", itqr, itpo);
+    }
+
+    if (itqr + itpo > 6) {
+        printf("\nConverged after %d. Check what is the issue, "
+                   "because QDWH needs <= 6 iterations.\n",
+                   itqr+itpo);
+    }
+
+    // A = U*H ==> H = U'*A ==> H = 0.5*(H'+H)
+    auto AT = conj_transpose(A);
+    if (m == n) {
+        auto W10_n = W1.slice(0, n-1, 0, n-1);
+        gemm(one, AT, Acpy, zero, W10_n, opts);
+        slate::copy(W10_n, H, opts);
+    }
+    else {
+        gemm(one, AT, Acpy, zero, H, opts);
+    }
+    // todo: try something like her2k to compute H
+    //her2k(one, A, W10, rzero, H, opts);
+    //auto AL = HermitianMatrix<scalar_t>(
+    //        slate::Uplo::Lower, H );
+    //slate::copy(AL, H, opts);
+}
+
+} // namespace impl
+
+//------------------------------------------------------------------------------
+/// Distributed parallel polar decomposition based on QDWH algorithm.
+///
+/// Computes a polar decomposition of an m-by-n matrix $A$.
+/// The factorization has the form
+/// \[
+///     A = UH,
+/// \]
+/// where $U$ is a matrix with orthonormal columns and $H$ is a Hermition
+/// matrix.
+///
+/// Complexity (in real):
+/// - for $m \ge n$, $\approx 2 m n^{2} - \frac{2}{3} n^{3}$ flops;
+/// - for $m \le n$, $\approx 2 m^{2} n - \frac{2}{3} m^{3}$ flops;
+/// - for $m = n$,   $\approx \frac{4}{3} n^{3}$ flops.
+/// .
+//------------------------------------------------------------------------------
+/// @tparam scalar_t
+///     One of float, double, std::complex<float>, std::complex<double>.
+//------------------------------------------------------------------------------
+/// @param[in,out] A
+///     On entry, the m-by-n matrix $A$.
+///     On exit, if return value = 0, the orthogonal polar factor $U$
+///     from the decomposition $A = U H$.
+///
+/// @param[out] H
+///     On exit, if return value = 0, the hermitian polar factor matrix $H$
+///     from the decomposition $A = U H$.
```

Comment:
Is this `B`? It must be named the same as in the arguments.

Document `itqr` and `itpo`.